File: app/src/main/java/com/merino/ddfilms/api/TMDBClient.java
package com.merino.ddfilms.api;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class TMDBClient {
    private static final String BASE_URL = "https://api.themoviedb.org/3/";

    private static Retrofit retrofit;

    public static Retrofit getClient(String apiKey) {
        if (retrofit == null) {
            OkHttpClient client = new OkHttpClient.Builder().addInterceptor(chain -> {
                Request original = chain.request();
                Request.Builder requestBuilder = original.newBuilder().header("Authorization", "Bearer " + apiKey).header("Content-Type", "application/json").header("accept", "application/json").method(original.method(), original.body());

                return chain.proceed(requestBuilder.build());
            }).build();

            retrofit = new Retrofit.Builder().baseUrl(BASE_URL).client(client).addConverterFactory(GsonConverterFactory.create()).build();
        }
        return retrofit;
    }
}

File: app/src/main/java/com/merino/ddfilms/api/TMDBClient.kt
package com.merino.ddfilms.api

import okhttp3.OkHttpClient
import okhttp3.Request
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object TMDBClient {
    private const val BASE_URL = "https://api.themoviedb.org/3/"
    private var retrofit: Retrofit? = null

    @JvmStatic
    fun getClient(apiKey: String): Retrofit {
        if (retrofit == null) {
            val client = OkHttpClient.Builder().addInterceptor { chain ->
                val original = chain.request()
                val requestBuilder = original.newBuilder()
                    .header("Authorization", "Bearer $apiKey")
                    .header("Content-Type", "application/json")
                    .header("accept", "application/json")
                    .method(original.method, original.body)

                chain.proceed(requestBuilder.build())
            }.build()

            retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        }
        return retrofit!!
    }
}
